### ISSUE TITLE
`induced_slot` mutates values and does a hidden `deepcopy` and generators have become reliant on that

### ISSUE THREAD

**Issue by sneakers-the-rat** [linkml/linkml#2219]
**Describe the bug**

this is a pretty whacky one!!!! has taken me about 8 hours to wrap my head around...

So i was out perfbusting, right. and `induced_slot` was as usual, the culprit. 

## Act 1: `required`

So first I noticed that attributes are set on the wrong object here, on the `slot` class rather than the `induced_slot` - that's bad right? we're mutating the parent slot that we copied from, defeating the purpose of the copy:

https://github.com/linkml/linkml-runtime/blob/984e4b2657f681dfa50e4cbd692bdb946ea56976/linkml_runtime/utils/schemaview.py#L1380-L1383

So i swapped that out with `induced_slot` in both places and got an error in this test: https://github.com/linkml/linkml/blob/bd49d48575d20e80875c2e1ef735d910f3a0d250/tests/test_compliance/test_core_compliance.py#L505

for the SQL generator, the `id` slot is no longer required. This fails in a few different tests for a few generators - the SQL generators and the SHACL generators. 

i was like "oop, that's definitely a bug," because we shouldn't be depending on accidental mutation of the parent `slot` object, and started opening this issue.

But i wanted to have a working version of the method in hand, so i kept working and things got stranger...

## Act 2: `slot_usage`

The main thing i wanted to tackle `induced_slot` doing this iteration over `(n fields in SlotDefinition) * (n parent classes)` here: **https://github.com/linkml/linkml-runtime/blob/984e4b2657f681dfa50e4cbd692bdb946ea56976/linkml_runtime/utils/schemaview.py#L1352-L1374**

I figured "hey all that's doing is grabbing `slot_usage` and updating from it, so why don't get just iterate over `class_ancestors` and only do stuff if `slot_usage` is declared for that ancestor and cut out like 99% of the iteration time." 

So I refactored it to do so and then got a bug that i couldn't explain at all - in `test_slot_usage` ( https://github.com/linkml/linkml/blob/bd49d48575d20e80875c2e1ef735d910f3a0d250/tests/test_compliance/test_inheritance_compliance.py#L456 ) the SQL generator was making an SQLA model like...

(in pseudocode bc i can't remember how i got it to display the generated SQLA as i'm writing this)

```python
class D():
    s1_id: 'Y' = ForeignKey('X.id')
    relations = Relation('Y')
```

or, in prose, the schema was correctly setting the `range` of the `SlotDefinition` as `Y`, but the `SlotDefinition.annotations['s1_id']['foreign_key']` was instead pointing to `X`.

What the heck!?!?!?!

IF WE READ THE FLOW CONTROL HERE CLOSELY, there is a special case that makes `slot_usage` work, but it's not what you think it is. 

`anc_slot_usage` here is a `SlotDefinition` - https://github.com/linkml/linkml-runtime/blob/984e4b2657f681dfa50e4cbd692bdb946ea56976/linkml_runtime/utils/schemaview.py#L1363

which creates empty `{}` and `[]` objects for multivalued metaslots. SO `v` earlier and the next line getting `v2` will not be `None` for many of the fields like `annotations`, etc. CRUCIALLY the `jsonasobj2.__setattr__` method will CREATE A NEW JsonObj here: https://github.com/linkml/linkml-runtime/blob/984e4b2657f681dfa50e4cbd692bdb946ea56976/linkml_runtime/utils/schemaview.py#L1379 when given an empty dict: https://github.com/hsolbrig/jsonasobj2/blob/51d966a97bc123561f31f06480957ba8a14f71af/jsonasobj2/_jsonobj.py#L163 

so this is why `induced_slot` is still so expensive - for every non-scalar metaslot for every class ancestor for every slot for every class, a new JsonObj is being created.

AND that explains the problem with our mismatch between `range` (a scalar metaslot) and `annotations` which carries the foreign key information: it only ever worked by accident, because the `annotations` dict in the implied `SlotDefinition` in the `slot_usage` object was empty, we created a new `annotations` object in the child, where otherwise it would share the same `id` as the parent class's attribute. 

This is not a problem in most of the generators because they don't rely on mutating the values that come out of `SchemaView`, but the `RelationalModelTransformer` used by the SQL generators does - see here where two schemaviews are mixed and lots of elements are created from parts of the other, mutated, etc. So the "view" nature of schemaview is violated by relying on that mutation. 

## Conclusion

SO there are a few things here - i would view this behavior as a bug, and we're sorta forced into choosing between a few bad options, mostly because `jsonasobj2` makes doing some basic things like "only getting the fields that are set in `slot_usage`", creating new objects, and even testing for truthiness very expensive (if you switch the `if obj` to `if obj is not None` then you save something like 20% of the runtime of `induced_slot` because `__bool__` is slow).  It shouldn't be that expensive to create/deepcopy a new `SlotDefinition` when it only has two non-empty fields, but it is, so the bandaid fix of re-adding the `deepcopy` into `induced_slot` makes us take back a significant perf penalty.

basically imo there shouldn't be any alternative in schemaview to `induced_slots` - if you want the literal representation of the slot definition, you can load the `SchemaDefinition` from the yaml and introspect from there, but `SchemaView` should always give fully hydrated schema elements. to make that efficient, we need to make our container objects like `SlotDefinition` more nimble, but also [make](https://github.com/linkml/linkml/issues/1839) [schemaview](https://github.com/orgs/linkml/discussions/1739#discussioncomment-8226761) [recursive](https://github.com/orgs/linkml/discussions/1739#discussioncomment-8600477) so that eg. in `induced_slots` we are only ever considering the immediate parents of the slot, and with imports we are only ever considering the first-layer imports, and thus our caching can be smarter because we have *more* objects but their creation is more deterministic from having a more limited scope. that would also let us greatly simplify schemaview so there are fewer opportunities for mysterious bugs :)

anyway for now i'm gonna just call this a bug, submit the perf improvements that don't break anything, and monkeypatch around the rest bc i don't use the SQL generators and probably won't have time to fix the problems, but imo they are relatively important bugs to fix bc eg. it's only because there are explicit tests for `slot_usage` that show the implicit mutation, but there are probably other places where inheritance is going wonky from this.

Related to: 
- https://github.com/linkml/linkml/issues/2217
- https://github.com/linkml/linkml/issues/1966
- https://github.com/linkml/linkml/issues/1665
- https://github.com/linkml/linkml/issues/971
- https://github.com/linkml/linkml/issues/1604
- https://github.com/linkml/linkml-runtime/pull/296

**Comment by sneakers-the-rat** [comment]
fwiw my attempt at perf improvement here: https://github.com/sneakers-the-rat/linkml-runtime/tree/perf-induced-slot

ends up being faster than the current implementation even with the deepcopy added in (which takes 73.7/87.6=84% of the time, and the rest of that time is because I have to instantiate a `SlotDefinition` with the `slot_usage`s to cast things to their proper types) when run on the `slow` tests:

<img width="791" alt="Screenshot 2024-07-23 at 1 11 10 AM" src="https://github.com/user-attachments/assets/b36935df-3102-4ab3-b039-5f92b5238391">

compared to the current implementation:

<img width="807" alt="Screenshot 2024-07-23 at 1 12 13 AM" src="https://github.com/user-attachments/assets/cc8176ac-7352-421a-b2c4-c7045cf6952d">

because
- the iteration over every metaslot for every parent class is really expensive (big gray bar on the left)
- `Schemaview.__hash__` is recomputed roughly a billion times (tan bar), see https://github.com/linkml/linkml-runtime/pull/329
- `JsonObj.__bool__` is slow (green)
- `JsonObj.__setattr__` is slow (light blue)

**Comment by sneakers-the-rat** [comment]
[this version](https://github.com/sneakers-the-rat/linkml-runtime/commit/657cde35caa05b6ed4a078b9d68045cc3c9f194c) is faster (run over the metamodel tests, same as above) by avoiding needing to make a bunch of empty objects with deepcopy, but now `remove_empty_objects` is the costly call
<img width="907" alt="Screenshot 2024-07-23 at 1 56 01 AM" src="https://github.com/user-attachments/assets/a3d19f9a-8899-4be7-bd37-afb9f7dc5cde">
. 

and finally [this version](https://github.com/sneakers-the-rat/linkml-runtime/commit/c22ca17161547378fde13f26133eece6b3ed2cfc) is the fastest i can get it with the tests still passing, which is basically what i think what we wanted to do there anyway
<img width="816" alt="Screenshot 2024-07-23 at 2 01 44 AM" src="https://github.com/user-attachments/assets/069c8384-217a-48d8-adb0-b5a39b0cf25c">

recall that's just over the metamodel tests, and the difference is huge when i'm generating nwb in realtime - the difference between a snappy and responsive tool you would want to use vs one that keeps ya waiting <3

**Comment by nlharris** [comment]
Is there still work to be done on this?

**Comment by sneakers-the-rat** [comment]
yes - we are still mutating the base slot rather than the induced slot: https://github.com/linkml/linkml-runtime/blob/ad74966db72be214a38de19f361df88489286e83/linkml_runtime/utils/schemaview.py#L1410-L1413

and it still breaks places that have come to rely on the implicit mutation of the slot object, observable with `poetry run python -m pytest tests/test_compliance/test_core_compliance.py -k "test_cardinality"` with a version of `linkml-runtime` that has the above lined swapping `slot` with `induced_slot`
